### ISSUE TITLE
Netatmo plugin -- Auth update  May 29, 2024

### ISSUE THREAD

**Issue by Phil353556** [domoticz/domoticz#6036]
Hi, today everything is working perfectly with Netatmo, thank you so much.
However as a netatmo customer, I just received this mail, I copy paste, fyi,  I don't know if it is already taken into account in the plugin or not. Thank you.

----
Authentication update (second notice)

This is a reminder regarding our authentication update following the release initially planned for last fall which we decided to postpone in order to give our users more time.
 
Dear Netatmo developer,

Today, when you refresh an Access Token using the associated endpoint https://api.netatmo.com/oauth2/token, the Netatmo servers respond with a pair of tokens: an Access Token and a Refresh Token.

If the previous Access Token is still valid, the newly returned access token is identical but has an expiration time extended by 3 hours.

The Refresh Token is not renewed.

Starting May 29, 2024, this behavior will change to comply with the [OAuth2 Authorization Framework RFC recommendations (section 10.4)] (https://www.rfc-editor.org/rfc/rfc6749) and improve the security of our users' data.
When refreshing the tokens, the Access Token and Refresh Token will be different from the previous ones and the old tokens will be invalidated.

What does this mean for you?

If you already store the tokens returned by the API when refreshing your tokens, this change will not affect you.
If you do not update the tokens when refreshing them, your users will be logged out because the old tokens will be invalidated.
If for a given user, you have stored the same token in different places (for example: an application and its widget), you must have a way to synchronize them because the new token will be applied for the entire application.

Sincerely,

Legrand - Netatmo - Bticino
----

**Comment by Mrhugues** [comment]
Hi, every body.

Same question , all run fine , but at 29/05/2024 what happen ?

**Comment by waltervl** [comment]
Someone seems to work on this already: https://github.com/domoticz/domoticz/pull/6016

**Comment by waltervl** [comment]
Should already be implemented by https://github.com/domoticz/domoticz/pull/5815

**Comment by Phil353556** [comment]
Hi @waltervl 
Ok,
"When refreshing the tokens, the Access Token and Refresh Token will be different from the previous ones and **the old tokens will be invalidated**."
will see soon :-)
Thank you.

**Comment by Salan54** [comment]
@Phil353556  It appears that your predictions were accurate. After following the method outlined on the Domoticz wiki to update my Netatmo configuration, everything seems to be functioning correctly for a few hours. However, I've noticed an error appearing in my logs around this time:

2024-06-21 17:16:10.514 Error: Netatmo: No access granted, forcing login again...
2024-06-21 17:16:10.775 Error: Netatmo: No access granted, check credentials...

After experiencing this issue, I restart the wiki procedure from scratch and it functions correctly for a few hours once again.

Am I the only one who has encountered this problem?

**Comment by JurgenLB** [comment]
You only need to renew the return code ('password in Domoticz') so no need to change the clientID / client secret.

For me I need to renew the 'password' when I restart Domoticz.

Kind regards

**Comment by Phil353556** [comment]
Hi, 
Agree,
Each time domoticz is restarted, yes, you need to renew the "password"

@Salan54 thanks, this is not prediction, but what Netatmo wrote to their users. You don't receive a mail from them?

Kind regards

**Comment by Phil353556** [comment]
Hey, have a look here: https://dev.netatmo.com/apidocumentation/oauth

And to netatmo communiy:
https://helpcenter.netatmo.com/hc/en-us/community/posts/19506310228242-API-token-invalid-after-3-hours-despite-refresh

_"Now, the access/refresh_token values are changed at each new call to /token endpoint and former values invalidated. So, after 3 hours when it's time to refresh your access_token, you must check in your code that you correctly stored the refresh_token value

Finally, use it in your refresh_token process, get a new pair of access/refresh_token values in the JSON response, and store the refresh_token for the next /token request"_

**Comment by jcbraem** [comment]
Hello,
One automatic solution is considered with Domoticz ?
Thanks,

Jean-Charles

**Comment by Salan54** [comment]
Hello everyone,

I'd like to share my experience with the Netatmo plugin on Domoticz. Despite receiving the necessary email notification from Netatmo in a timely manner, I didn't initially realize that it would affect my Domoticz setup - silly mistake, indeed!

In my case, I found that I needed to update the password every few hours, not just upon each Domoticz restart as I initially thought. It would be incredibly helpful if the plugin could automatically refresh the access token, eliminating the need for manual updates.

That being said, I do appreciate your help and assistance.

Best regards,

**Comment by geschriever** [comment]
Hi all, yes I encountered the problem also and a new password will last only a few hours. If there can be a solution this would be great.

**Comment by Phil353556** [comment]
Hi,
I use **domoticz 2024.4**  on a:
-Raspberry Pi 3 Model B Rev 1.2
-Debian GNU/Linux 11 (bullseye) aarch64

The last time I renew the "password" was, when domoticz has been restarted on june 21.
Since june 21 all measurements from netamo weather / rain  were collected, correctly, with no problem.

Please note, only this URL is working, do not try another one with other scope
https://api.netatmo.com/oauth2/authorize?client_id=xxxxxxxxxxxxxxxxxxxxxx&redirect_uri=http%3A%2F%2Flocalhost%2Fnetatmo&state=teststate&scope=read_station%20read_thermostat%20write_thermostat%20read_homecoach%20read_smokedetector%20read_presence%20read_camera

hope this helps
Phil

**Comment by Chrominator** [comment]
I've upgraded to version 2024.4 (build 16089) to be sure to use the change made for May 29th, 2024.
As others, every domoticz restart requires a new password.
Worked fine for several days, this is the log for Netatmo at restart: 
```
2024-06-25 20:05:27.978 Error: Netatmo: No access granted, forcing login again...
2024-06-25 20:05:28.155 Error: Netatmo: No access granted, check credentials...
2024-06-25 20:05:58.316 Error: Netatmo: No access granted, forcing login again...
2024-06-25 20:05:58.459 Error: Netatmo: No access granted, check credentials...
```

**Comment by Phil353556** [comment]
Hi,
I think we have to share experiences, to determine the number of days before the issue is up.
Following a power outage:

024-06-26 08:40:39.997 Error: Netatmo: No access granted, forcing login again...
2024-06-26 08:40:40.350 Error: Netatmo: No access granted, check credentials...
2024-06-26 08:40:58.741 Status: [web:8080] Incoming connection from: 192.168.6.18  ->>>> password changed
2024-06-26 08:41:04.797 Status: Netatmo: Worker stopped...
2024-06-26 08:41:04.798 Status: Netatmo: Worker started...
2024-06-26 08:46:15.052 Status: Netatmo: cloud data for module [rain gauge] is now updated again
2024-06-26 08:46:15.053 Status: Netatmo: cloud data for module [External Module] is now updated again

**Comment by JurgenLB** [comment]
I think it is because off the hardcoded fixed scope.

We must let users chose the scope en authenticate with that scope.

Kind regards

**Comment by Phil353556** [comment]
I think, you're right. I tested without the scope "smoke_detector" as I don't have one, it doesn't work.
So I use the URL as writtten even if I don't have a smoke detector :-)

**Comment by track53** [comment]
Hi all,
I have a script with netatmo that mesur internal / external temp and if to hot that close my blind. But I see my netatmo no refresh tan 24/12/2023. And I have read a lot of post with new authentification for netatmo, I have create my apps , I have the client ID and secret ID, I launch the url and I copy the code was generate but Domoticz will not connect to netatmo.

 2024-06-27 09:16:21.695 Error: Netatmo: No access granted, check credentials...
2024-06-27 09:16:52.018 Error: Netatmo: No access granted, check credentials...
2024-06-27 09:17:22.387 Error: Netatmo: No access granted, check credentials...
2024-06-27 09:17:52.777 Error: Netatmo: No access granted, check credentials...
2024-06-27 09:18:23.142 Error: Netatmo: No access granted, check credentials... 

Or 

 2024-06-27 08:44:54.561 Error: Netatmo: Error connecting to Server...
2024-06-27 08:44:54.930 Error: Netatmo: Error connecting to Server...
2024-06-27 08:59:55.415 Error: Netatmo: Error connecting to Server...
2024-06-27 08:59:55.778 Error: Netatmo: Error connecting to Server... 

I am on domoticz 2024.4 can you help me ?

**Comment by Phil353556** [comment]
Hi,
I think I have read that in this case, you should have two differents applications (api) with netatmo.
One as described above for domoticz
Second , you create for your application
Otherwise, if only one application, that means that following each call to the api must update the token for both ,  domoticz and your application. I don't know how to interact with domoticz.
regards
Phil

**Comment by track53** [comment]
Hi Phil353556, exactly as same time I try with the first apps CLient ID .... and now it's working.
I think you can add this step in the wiki if An apps already exist we will use this client ID and secret ID.

**Comment by Phil353556** [comment]
Hi @track53 
I doubt, when I read you.
Do I understand, that you use exactly the same  information for both applications?  ie  for your script AND for  domoticz/netatmo ?
    the username should contain **<client_id>:<client_secret>**
    the password should contain the retrieved   <**SOMECODE**>
    Thank you for your answer
    Phil

**Comment by track53** [comment]
Hi @Phil353556,
Yes I have already create an application netatmo Legrand to try open/close my solar bubendorf, and this week I have create new one for the new authentification with Domoticz I create new client ID ..... not work. I have try to use the secret .... of my first app and that work. Sorry for my english.

**Comment by JurgenLB** [comment]
Hi @track53 the bubendorff scopes are not currently supported with the hardcoded fixed scope,
I'm working on the new API but the callback from the selector-switch are giving me some trouble.
 I don't known witch devices are in the budendorff-scope
These are the different scopes possible;
[Netatmo scopes.pdf](https://github.com/user-attachments/files/16039580/Netatmo.scopes.pdf)

I can share my latest work but its not yet finished.


Kind Regards

Jurgen

**Comment by Chrominator** [comment]
Enter new password after a restart on July 9th, 20:30.
No updates after July 11th, 14:48 without any restart since entering the password.
It is not the first time, seems the tokens are not renewed after a while, nothing in the log about invalid credentials...
Very strange to me.

Manual restart of Netatmo hardware, new message in the log:
2024-07-11 19:15:14.354 Status: Netatmo: Worker stopped...
2024-07-11 19:15:14.354 Status: Netatmo: Worker started...
2024-07-11 19:15:19.560 Error: Netatmo: No access granted, check credentials...

**Comment by Phil353556** [comment]
@Chrominator 
strange it is 
still using domoticz 2024.4,  I didn't update yet to the latest 2024.6, and I think I will not do before september
domoticz started on june 29, no problem until now
Is there somewhere a detailed log for the plugin? @gizmocuz 
```
ps -ef    |   grep domo
avahi        399       1  0 Jun29 ?        00:04:26 avahi-daemon: running [domoticz.local]
root        1350       1  8 Jun29 ?        1-01:10:59 /home/domoticz/domoticz/domoticz -daemon -www 8080 -sslwww 443 -log /tmp/domoticz.txt
```

from time to time these messages, sometime only one sensor, sometimes two, sometimes three , 
total is 51 messages since june 29
Tranlation:
-Extérieur ->  external sensor
-Pluviomètre -> external rain gauge
-Intérrieur -> Internal sensor
```
2024-07-10 20:07:46.379  Error: Netatmo: cloud data for module [Extérieur] no longer updated (module possibly disconnected)
2024-07-10 20:07:46.379  Error: Netatmo: cloud data for module [Pluviomètre] no longer updated (module possibly disconnected)
2024-07-10 20:07:46.380  Error: Netatmo: cloud data for module [Intérieur] no longer updated (module possibly disconnected)
....
2024-07-10 20:37:47.879  Status: Netatmo: cloud data for module [Extérieur] is now updated again
2024-07-10 20:37:47.880  Status: Netatmo: cloud data for module [Pluviomètre] is now updated again
2024-07-10 20:37:47.880  Status: Netatmo: cloud data for module [Intérieur] is now updated again
```

**Comment by LoukKamp** [comment]
Same trouble here (needed to generate a new token after each restart lately), but since yesterday, I can't get it working again.

Could there be any changes in the new release of Domoticz (2024.6) released yesterday? Note, I'm using beta releases too.

New for me are the errors about connecting to the server:


` 2024-07-11 23:56:50.788 Status: NetAtmoWeerEnLuchtkwaliteit: Worker stopped...
2024-07-11 23:56:50.788 Status: NetAtmoWeerEnLuchtkwaliteit: Worker started...
2024-07-11 23:56:56.205 Error: NetAtmoWeerEnLuchtkwaliteit: **Error connecting to Server...**
2024-07-11 23:56:56.400 Error: NetAtmoWeerEnLuchtkwaliteit: **Error connecting to Server...**

(restarted the plugin)
2024-07-11 23:57:26.535 Status: NetAtmoWeerEnLuchtkwaliteit: Worker stopped...
2024-07-11 23:57:26.536 Status: NetAtmoWeerEnLuchtkwaliteit: Worker started...
2024-07-11 23:57:31.731 Error: NetAtmoWeerEnLuchtkwaliteit: No access granted, check credentials...
2024-07-11 23:58:01.983 Error: NetAtmoWeerEnLuchtkwaliteit: No access granted, check credentials...
2024-07-11 23:58:32.190 Error: NetAtmoWeerEnLuchtkwaliteit: No access granted, check credentials...
2024-07-11 23:59:02.399 Error: NetAtmoWeerEnLuchtkwaliteit: No access granted, check credentials...
2024-07-11 23:59:32.595 Error: NetAtmoWeerEnLuchtkwaliteit: No access granted, check credentials... 
`

**Comment by Phil353556** [comment]
Hi, each time the plugin is restarted,  it is needed to get back with a new password following the usual procedure
I don't know if it is a bug or not @gizmocuz may tell us?

**Comment by LoukKamp** [comment]
I agree, but this was working until I did the update to the latest release last Thursday. Now I get this "Error connecting to Server..." messages after retaining a new code and restarting the worker.

After some digging in debug info, I found that this is because the access to the requested page is forbidden:

`
2024-07-12 23:27:02.053  [ef4ff220] Status: NetAtmoWeerEnLuchtkwaliteit: Worker stopped...
2024-07-12 23:27:02.054  [ef4ff220] Status: NetAtmoWeerEnLuchtkwaliteit: Worker started...

2024-07-12 23:27:07.549  [ef4ff220] Debug: **HTTP 403: Forbidden**
2024-07-12 23:27:07.554  [ef4ff220] Error: NetAtmoWeerEnLuchtkwaliteit: Error connecting to Server...
2024-07-12 23:27:07.778  [ef4ff220] Debug: **HTTP 403: Forbidden**
2024-07-12 23:27:07.785  [ef4ff220] Error: NetAtmoWeerEnLuchtkwaliteit: Error connecting to Server...
`

When I change anything to the device or restart Domoticz, the worker will be restarted, and then it will try to get a new token using the code I retrieved from NetAtmo before:

`2024-07-12 23:23:39.392  [ef4ff220] Debug: NetAtmoWeerEnLuchtkwaliteit: No access granted, check credentials...
  (
    grant_type=authorization_code
    &client_id=<my client id>
    &client_secret=<my client secret>
    &code=57df735b9804f7dc29ed82cd02caa8c0
    &redirect_uri=http://localhost/netatmo
    &scope=read_station read_thermostat write_thermostat read_homecoach read_smokedetector read_presence read_camera
  )
 
 (
    {
        "access_token" : null,
        "error" : "**invalid_grant**"
    }
  )
`
Note the this code was only valid for some seconds, so retrieving a new toke based on this code won't ever work.

I don't know why this is done every time time the worker or Domiticz is restarted., because there is a valid token available in the database (table Hardware, field Extra). I'm also not shure if this is the access token (valid a view hours?) or the refresh token (valid for several days?).  Why don't use this token to restart the sequence after a short time? When both tokens are expired there is no use in keep trying. In this case, just notify the user that (s)he should request for a new code and stop the worker.

I hope this will help one of you guys to find a proper solution for this issue.

**Comment by LoukKamp** [comment]
Tonight, I downgraded to V2024.4 by replacing the domoticz image and changing the db version from 169 back to 167 (yes, I know this is a risky thing to do, but I made a backup ;)).

However, this did not change the behavior of the Netatmo module: Still getting this "Error connecting to Server..." messages after I replaced the authorization code with a new one (retrieved from the Netatmo API site). So I'm still unable to refresh my access and refresh tokens. 

After restarting the worker, it keeps trying to get new tokens using the authorization code that is already expired.

So my suggestion that something has been changed in domoticz since the last stable release was wrong! However, something must have been changed since last week Thursday, because, before that time I was able to get new valid tokens by refreshing the authorization code (required after each restart of the worker).

@JurgenLB : Jurgen, can you **please** help my out, by supplying me an image that has the debug lines you have added in your project, so I can see what is actually happening and track down this issue?

**Comment by JurgenLB** [comment]
@LoukKamp It is my experience that if you use a scope that is different from the devices that's in your Netatmo account. It is Netatmo who rejects your authentication.
This is because Netatmo has changed their API and authentication. This is not a change in Domoticz.
It is true that the plugin in Domoticz need to be adapted to meet the new authentication method from Netatmo.

In my project I'm adapting to the new API for compatibility with all devices, the authentication is a bigger change in Domoticz.

I have this experience because in my project I added the scope for the keylock because I wanted to purchase this at a later moment, but I don't have it in at this moment. And after a while my authentication was blocked by Netatmo with the same message your referring to.
It is true that because off my debugging lines I better understand the error.

If you can compile the Domoticz on your local machine you can put the debugging lines and the correct scope for your situation in the plugin-code. And than with a new registration (other email account) you can use the plugin.

But this is a temporarily  workaround and not doable for everybody.

Hope you can work with this

Kind regards

**Comment by LoukKamp** [comment]
Thanks for the quick reply. However, I was aware of the need to use this fixed scope and did use it.

After double-checking this scope, user ID, and password, I renewed the User ID and password for the app by resetting the keys: Still the same trouble.

Finally, I created a new app on the NetAtmo site. The last action was finally successful, and I was able to get things working again :) !!!

Although I'm very happy that I'm able to access my own data again, I still don't see why the old app didn't work anymore: The app was enabled, and the state was OK.

**Comment by LoukKamp** [comment]
But this still leaves us with the issue of trying to restart the token renew process by using the expired authorization code. 

I'm willing to help to tackle this, but I don't have a development environment set up on this (or any other) Linux system. This makes debugging quite a hell...

So, if anyone can help my with an image that includes the required debugging lines (like the thread Jurgen is working on), or with a link to some SIMPLE guide to set up a development environment to build my own, I would be very grateful.

Note: I'm not an experienced Linux / C++ developer, although I have quite some experience in developing and troubleshooting in a Windows / C# environment.

**Comment by Chrominator** [comment]
FYI, no update after several days/hours but no error message in the log:

` grep "Netatmo:" /var/log/domoticz.log > /tmp/netatmo.log`
[netatmo.log](https://github.com/user-attachments/files/16325856/netatmo.log)

16:16 nothing special appended, no further update.
Worker seems just stop to update the devices.

Snapshot taken at 22:02 July 21st.

**Comment by LoukKamp** [comment]
But this still leaves us with the issue of trying to restart the token renew process by using the expired authorization code. 

I'm willing to help to tackle this, but I don't have a development environment set up on this (or any other) Linux system. This makes debugging quite a hell...

So, if anyone can help my with an image that includes the required debugging lines (like the thread Jurgen is working on), or with a link to some SIMPLE guide to set up a development environment to build my own, I would be very grateful.

Note: I'm not an experienced Linux / C++ developer, although I have quite some experience in developing and troubleshooting in a Windows / C# environment.

**Comment by LoukKamp** [comment]
Edit: I forgot about the 3rd issue; added this to the list below.

I think we have 3 issues here:

1.  When the worker is restarted for some reason, it will try to use the refresh token to renew the access token and refresh token. When this fails (see below), it will try to log in again, using the authorization code. However, this code is only valid for several seconds, so this request will always fail because this code is outdated. So we might need another process to get a new authorization code (if required).

2. When the worker is started and the user has provided a new authorization code (manually by entering this code in the password field of the device), everything will work fine until the worker is restarted for some reason. However, after 3 hours, the access token will expire and needs to be renewed. The worker will post a request to renew the tokens, and uses these codes to acquire data and every 3 hours a new set of tokens. However, those tokens **are not stored in the database**!!! So, when the worker is restarted, it will use the tokens received after providing the authorization code, and they are not valid anymore...

3. When the worker is restarted (without restarting Domoticz), the first post is a login request using a "grant_type=authorization_code" with the expired authorization code instead of trying to use the refresh token that might still be valid. This **might** be caused by the lag of initialization of the variable _m_isLogged_  ,  _m_nextRefreshTs_, or _m_refreshToken_ . I'm still trying to figure this out.

The cause of the second issue lies in the change Netatmo applied some time ago (May 24, 2024): Before this change, the access and refresh tokens did not change when a refresh token request was sent before the tokens expired. But after this change, a new set of tokens are generated and the previous tokens are invalidated. 

So, now we need a change that stores the refresh token every time a token refresh call is made, so the worker will use the latest valid refresh token to refresh the tokens at startup.. 

As I'm unable to test this myself (don't have a Linux development environment at this moment), can someone else (maybe @JurgenLB ?) please verify this theory, make a code change, test this change, and make this available in the next beta release?

**Comment by Phil353556** [comment]
Hi,
My 2cts,  version domoticz 2024.4 (January 2024)  restarted july16, new netatmo password put on the same day,  still up and running without any problem.
I DO not update domoticz to 2024.6
kind regards

**Comment by LoukKamp** [comment]
I tried that (down grading to 2024.4), but didn't notice any improvements. And as far as I know, the code for the Netatmo thread didn't change since the last year(s?).

If I understand your post well, you just started Domoticz, entered a new authorization code in the password field, and pressed the update button.

You didn't restart Domoticz or changed any settings for the Netatmo device (e.g. enable or disable a log option)?

What happens when you change one of those settings for this device, or even restart Domoticz?

**Comment by JurgenLB** [comment]
@Phil353556  can you share with devices you have ?
Because if the scope is correct for you, the refresh keeps working.

Maybe we can use the NetatmoScope as password in domoticz and automate the return code but then we still must have a pop-up for the user to accept the authentication.

It is my believe that the users who have issues does not have all the devices corresponding the NetatmoScope in Domoticz

Maybe there can be a survey  on the Domoticz Forum to have users list their devices-types and result off the authentication when restarting.

**Comment by Phil353556** [comment]
Hi @JurgenLB @LoukKamp 
_"If I understand your post well, you just started Domoticz, entered a new authorization code in the password field, and pressed the update button."_
-> Yes absolutely.

_"You didn't restart Domoticz or changed any settings for the Netatmo device (e.g. enable or disable a log option)?"_
-> Yes absolutely.

If domoticz is restarted, a new password is needed.

Initialy in january I tried to adapt the scope in the URL to my configuration,  but it wasn't not working,  so I keep the URL as it was even if there is no smoke detector in my configuration.
URL used to obtain password as describe in the wiki [https://www.domoticz.com/wiki/Netatmo]:
https://api.netatmo.com/oauth2/authorize?client_id=xxxxxxxxxxxxxxxxxxx&redirect_uri=http%3A%2F%2Flocalhost%2Fnetatmo&state=teststate&scope=read_station%20read_thermostat%20write_thermostat%20read_homecoach%20read_smokedetector%20read_presence%20read_camera
Yes it runs since july 16th
![image](https://github.com/user-attachments/assets/27365ff1-81fa-4537-acf1-db45bbc8a707)

My devices for weather, temperature and so on
-the base station, inside home, so this is the internal module, and only one , there is no additionnal internal module,
-a rain gauge (outside home, of course)
-an external module
--> no smoke detector, no wind gauge
I also have an internal camera, and an external camera, but there is no interaction with domoticz.

I hope this may help
kind regards

**Comment by LoukKamp** [comment]
I don't think this has anything to do with the scope and de devices you actually use. As long as the scope is the same as mentioned in de Wiki (and Phill's post above).

I made a TCPdump trace and it shows exactly what is going wrong (posts with expired access codes an refresh tokens). 

The only thing I don't understand is what is going wrong  after a restart of Domoticz within 3 hours of entering a new password (authorization code). In this case the token refresh request should contain a valid refresh token. I still need te reproduce this situation and see why this first request is rejected.

Sadly enough, I'm away from home for vacation, so testing is a bit different...

**Comment by LoukKamp** [comment]
Sorry to border you again, but I ran another test with trace that confirmed my expectation: 

1.  Started Domoticz a terminal (so I could enable SSH logging)
2. Requested a new security code from Netatmo (logging in)
3. Entered this code as a new password
4. Examined the stored password and the refresh token in the database
5. Stopped and restarted Domoticz on the terminal
6. Checked if the Netatmo device was still updating the database with measured values: OK
7. Checked the database for the stored password and refresh token: Still the same!!! Note that the worker did a request for new tokens at startup, so the old values (still stored in the database) are not valid anymore!
8. Leave this running for 3 hours, so the worker will refresh the tokens again.

After this, I will check the code and the refresh token in the database and I bet that they are still the same. So, when I restart Domoticz another time, the request won't be granted, because the refresh token is not valid anymore.

So we definitely need a change that stores the refresh token in the database after a successful refresh. In addition, it would be nice when a refresh would be forced when the worker or Domoticz is stopped, so we have at least the full 3 hours to start it again without the need of a new authorization code.

I still wonder why this does work when Domoticz is restarted, and it doesn't work when just the worker is restarted (e.g. by changing a parameter for the device). In the last case, it doesn't try to refresh the tokens, but starts a login procedure (using the old expired authorization code) right away. I'm still trying to figure this out...

**Comment by Phil353556** [comment]
Hi,
@LoukKamp Congrats for your analyze
I didn't know the name of the person who wrote the plugin, but I am sure that @gizmocuz can help.

Another issue,  little one,  the value for the level of the battery of rain gauge is only taken at initial start.
So you know when domoticz and the plugin with the netatmo password starts :-) 
![Capture d’écran du 2024-07-30 13-52-23](https://github.com/user-attachments/assets/4d28da37-3f0b-4f1a-810f-9f08ce978aeb)

**Comment by Phil353556** [comment]
@LoukKamp 
the stuff should be there isn't it?
https://github.com/domoticz/domoticz/blob/e1d6faa6cb0f1660dad0311ffbbcd5b4c9abdc62/hardware/Netatmo.cpp#L230
function StoreRefreshToken

**Comment by LoukKamp** [comment]
Absolutely correct. Although I don't understand why you showed this specific line, but this is the file to be modified.

However, there might by some conflicts with the project @JurgenLB is working on in the same file. But, as this is a very small change, it's a small step for @JurgenLB to merge this with his changes, so there will be no conflicts when he is ready to apply his changes.

My proposal for this moment:

1. Add a call to StoreRefreshToken() after a successful renewal of the tokens in RefreshToken() before line 354.
2. Add a call to StoreRefreshToken() after the main worker loop in Do_Work() before line 199.

For the moment this will be all. However, I still want to figure out why this module doesn't call RefreshToken() after just a worker restart. Let me sleep on that. ;)

**Comment by Phil353556** [comment]
sorry for the line number, this is around .. functions RefreshToken,  StoreToken
I just look but didn't make a change.

**Comment by LoukKamp** [comment]
No problem; I already assumed that. Just wanted to be sure.

I wish I had enough experience with GitHub and the workflow within this community to apply this change, but I don't, so I sit on my hands...

So, If somebody is willing this to do for me (and the rest of the Netatmo users), please do!

B.t.w., I finished the test and after running for more than 3 hours, the fields in the database did not change, so this proves the need for this patch.

**Comment by JurgenLB** [comment]
@LoukKamp In my local setup I have put the StoreRefreshToken() in the function RefreshToken I will let this run and see if it makes a difference 

Kind regards

**Comment by Phil353556** [comment]
@LoukKamp 
congratulations.

I can do a PR in github with the changes you wrote:
_Add a call to StoreRefreshToken() after a successful renewal of the tokens in RefreshToken() before line 354.
Add a call to StoreRefreshToken() after the main worker loop in Do_Work() before line 199._

tell me, just to be sure:
-> StoreRefreshToken()  before line 354 so just before the line "return true;"
-> StoreRefreshToken() before line 199 so between line 198 "}" and line 199 "Log(LOG_STATUS, "Worker stopped...");" 

Is this correct?
Phil

**Comment by JurgenLB** [comment]
> @LoukKamp In my local setup I have put the StoreRefreshToken() in the function RefreshToken I will let this run and see if it makes a difference
> 
> Kind regards

after this I periodically inspect the Log and I noticed;
`18:33:11.831  Debug: Netatmo: Status 1
18:38:11.855  Debug: Netatmo: Weather 1

20:56:24.700  Debug: Netatmo: Status
21:11:24.779  Debug: Netatmo: Weather 1

21:11:56.093 Debug: Netatmo: Status 1
21:26:56.168  Debug: Netatmo: Weather 1

21:27:36.051  Debug: Netatmo: Status 1
21:33:21.083  Debug: Netatmo: Refresh https://api.netatmo.com/oauth2/token?
21:42:36.416  Debug: Netatmo: Weather 1

22:14:35.812  Debug: Netatmo: Status 1
22:29:35.891  Debug: Netatmo: Weather 1
`
So after my last restart the interval is approximately 5 min but later on the interval seems 15 min. 
This is with the suggested StoreRefreshToken() after a successful renewal of the tokens in RefreshToken()
I will put this in my PR thanks for the suggestion @LoukKamp 

kind regards

**Comment by LoukKamp** [comment]
@JurgenLB : Sorry, I don't really understand what is happening in your logging.  In the code, I see it is set to 900 seconds (15 minutes). Maybe this is the interaction with the commands you gave, like starting and supplying the code in the password field.

@Phil353556 :  We need to coordinate this with @JurgenLB : He is working on the same module. However, this is a large and complex change, it might take a while before this reaches the main or beta releases, I assume. My suggestion is to implement this (partial) authorization change as a separate PR that can be released as soon as possible.

In the meanwhile I succeeded to build my own test version of Domoticz and maybe this will lead to understanding and a solution of the second issue where it doesn't try to use the stored refresh token after a restart of the worker. So, maybe wait another day, so these two patches can be combined in one PR.

**Comment by LoukKamp** [comment]
> @LoukKamp congratulations.
> 
> I can do a PR in github with the changes you wrote: _Add a call to StoreRefreshToken() after a successful renewal of the tokens in RefreshToken() before line 354. Add a call to StoreRefreshToken() after the main worker loop in Do_Work() before line 199._
> 
> tell me, just to be sure: -> StoreRefreshToken() before line 354 so just before the line "return true;" -> StoreRefreshToken() before line 199 so between line 198 "}" and line 199 "Log(LOG_STATUS, "Worker stopped...");"
> 
> Is this correct? Phil

Sorry Phil, forgot to answer this, but yes, this is correct:
EDIT: changed the call to StoreRefreshToken() to RefreshToken(true) to force a request for new tokens, before exit, to allow the user for at least 3 hours to get it working again (without the need of entering a new authorization code in the password field).

             }
         }
          *RefreshToken(true);**

        Log(LOG_STATUS, "Worker stopped...");
}


        m_accessToken = root["access_token"].asString();
        m_refreshToken = root["refresh_token"].asString();
        int expires = root["expires_in"].asInt();
        //Store the duration of validity of the token
        m_nextRefreshTs = mytime(nullptr) + expires;

        **StoreRefreshToken();**
        return true;
}

**Comment by LoukKamp** [comment]
Very strange:  After cloning, rebuilding Domoticz from source, and applying the change as mentioned above, I can't reproduce the situation where the worker started a new login (i.s.o. trying to refresh the tokens first) after just a restart or the worker (e.g. after applying a change of the logging options); now this works as expected. Of course, this is good, but it doesn't help much in understanding what was going wrong. ;)

Here is a part of the logging with addition debug information:

![afbeelding](https://github.com/user-attachments/assets/e657bdc2-6813-4ff8-96f5-e1a72d49f056)

**Comment by JurgenLB** [comment]
The code is seperated in different blocks, I put in some debug lines to identify the active block.
Station Data parsed
Weather 1
Station Data parsed
HomeCoach 1
HomeStatus parsed
Events Data parsed
Status 1

so weather 1 means the code for weather station has reached his end and the login is successful.
Status 1 is the last block, weather is the first block.
so the timer is indeed 900 sec. 
but in the beginning there is a interval off 5 min.?

I only put this in the code;
`        //Store the duration of validity of the token
        m_nextRefreshTs = mytime(nullptr) + expires;
        StoreRefreshToken();
        return true;
}`

and it looks beter even a Stop / Start of domoticz service keeps the authentication going.

**Comment by LoukKamp** [comment]
Where did you insert this code?

Yes indeed, when the refreshed tokens are stored, it will survive a restart of the worker and even Domoticz itself.

The only case left is disabling the device and enabling it  after some time. At the moment you press the update button with the enable switch armed, the refresh token is cleared in the database, causing the login() procedure to start a login request (using the expired authorization code.

So, now the question is: Who and where???

It's _not_ in the Netatmo class. Right now,  I'm suspecting the webserver code and I'll try to add some extra debug lines in this module.

**Comment by JurgenLB** [comment]
I put the code in 'RefreshToken()' function.

The username and password are handled by the app, see `domoticz/www/app/hardware/Hardware.js`
But just a thought is it not logically that after a disabling the user must enter a new authentication ?

The agreement with Netatmo states that we must keep the authentication secret and not stored,
it also states we must keep the scope as low as possible. 

kind regards

**Comment by LoukKamp** [comment]
Ok, I see, you also just added this call to StoreRefreshToken(); so we agree on this ;)

As far as I can trace this, It seems that this is caused by /main/WebServerCmds.cpp: the procedure Cmd_UpdateHardware() stores the field extra in the database, but this field is empty. So, the next step is finding out why; maybe domoticz/www/app/hardware/Hardware.js will tell.

When I understand this well, we are not allowed to store the username and password of the Netatmo account, but is this also true for the tokens?

The second agreement could be a pain to implent, because we need some kind of selection list in the hardware page and a place to store it...

**Comment by JurgenLB** [comment]
No keeping the authentication a secret means in my understanding that it is not allowed to store the tokens . The username and password are already stored.

For the scopes I was thinking of a 3e text-field where the scope can be entered. 
The user must then correctly fill in the scope,
a selection list is ultimately better but to difficult I believe.

**Comment by LoukKamp** [comment]
> No keeping the authentication a secret means in my understanding that it is not allowed to store the tokens . The username and password are already stored.
> 
> For the scopes I was thinking of a 3e text-field where the scope can be entered. The user must then correctly fill in the scope, a selection list is ultimately better but to difficult I believe.

Strange: This would mean that on every restart of Demoticz, you need to enter the Client ID, Client secret, and the authorization code... If this true, I would stop immediately and sell all Netatmo stuff right away; I would not accept that as a user!

So, I think we are missing here something, because all other lookalike apps, like the apps on your phone, don't require such terrifying authorization methods. Maybe some security expert can tell us what to do here.?

I was thinking of storing it encrypted, but because Domoticz is open source, there is no way to hide the method and keys to anyone who is interested in my room temperatures.

About the scope, I think a free text field may be even more work to validate the user's input. This can be avoided by a selection box where you just need to select the items you need from a list. This should not be so hard to do for an experienced web designer. Those selection lists might even be a standard object in HTML 5. Sadly enough, the last time I did some (simple) frontend programming is at least 5 years ago, and I did that in Visual Studio C++.  I have no experience in JavaScript and Ajax at all, so when you need my help, don't expect rocket science within a few weeks ;).

Edit: See [this page](https://www.html.am/html-codes/forms/html-select-lists.cfm) for an example of such selection lists.

**Comment by LoukKamp** [comment]
Just made some small progress in tackling the authorization issue: It is definitely in the web interface: When the hardware page is opened, all parameters of all devices are read from the database. When the user clicks on the update button, almost all parameters of the device are updated in the database., also the field _extra_ that should hold the refresh token. However, this field is not filled by the underlying JavaScript modules, so it is empty. When the app is not disabled, the worker will store the token (locally stored in a variable) again in the database, just after the token has been cleared by the web server. So this is not very efficient., but in the end it is OK. The same happens when the user did disable the device and clicks the update button. But when the user enables the device again and clicks the update button, the web server will clear the token in the database and then starts the worker. In this case, the token is not locally stored, and will be retained from the database. Because there is no token in the database, the worker will post a request for new tokens, using the (expired) authorization code: End of story...

How to solve this? I think we have 3 options:

1.  Change the JavaScript (Hardware.js), so the field extra will be filled in when the data is stored back into the database.
2. Use another field (like _configuration_) i.s.o. _extra_ that is not stored by the web server.
3. Add another field to the database (e.g. _refreshToken)_ hardware table that is not handled by the web server.

The first option is the most logical option, but (for me) the most complex one.

**Comment by LoukKamp** [comment]
In addition to the first 3 suggestions, I have forth option that is even much simpler: Changing the code in main/WebServerCmds.cpp where the fields are stored in the database. There was already a specific situation for the RFX devices, where the extra field was excluded from the update (and insert for new devices), so I added the Netatmo type to the expression of if statement (from line 580) for the addition of new devices:

                       else if ((htype == HTYPE_RFXtrx433) || (htype == HTYPE_RFXtrx868) || (htype == HTYPE_Netatmo))
                        {
                                // No Extra field here, handled in CWebServer::SetRFXCOMMode and in CNetatmo::CNetatmo for Netatmo devices
                                m_sql.safe_query("INSERT INTO Hardware (Name, Enabled, Type, LogLevel, Address, Port, SerialPort, Username, Passwo>
                                        "DataTimeout) VALUES ('%q',%d, %d, %d,'%q',%d,'%q','%q','%q',%d,%d,%d,%d,%d,%d,%d)",
                                        name.c_str(), (senabled == "true") ? 1 : 0, htype, iLogLevelEnabled, address.c_str(), port, sport.c_str(),>
                                        mode2, mode3, mode4, mode5, mode6, iDataTimeout);
                                extra = "0";
                        }
                        else
                        {
 and for the update from line 701 (in the original code):

                                else if ((htype == HTYPE_RFXtrx433) || (htype == HTYPE_RFXtrx868) || (htype == HTYPE_Netatmo))
                                {
                                        // No Extra field here, handled in CWebServer::SetRFXCOMMode and CNetatmo::CNetatmo( for Netatmo
                                        m_sql.safe_query("UPDATE Hardware SET Name='%q', Enabled=%d, Type=%d, LogLevel=%d, Address='%q', Port=%d, >
                                                "Mode1=%d, Mode2=%d, Mode3=%d, Mode4=%d, Mode5=%d, Mode6=%d, DataTimeout=%d WHERE (ID == '%q')",
                                                name.c_str(), (bEnabled == true) ? 1 : 0, htype, iLogLevelEnabled, address.c_str(), port, sport.c_>
                                                mode1, mode2, mode3, mode4, mode5, mode6, iDataTimeout, idx.c_str());
                                        std::vector<std::vector<std::string>> result;
                                        result = m_sql.safe_query("SELECT Extra FROM Hardware WHERE ID=%q", idx.c_str());
                                        if (!result.empty())
                                                extra = result[0][0];
                                }
I implemented this change and it works as expected! :)

**Comment by JurgenLB** [comment]
> In addition to the first 3 suggestions, I have forth option that is even much simpler: Changing the code in main/WebServerCmds.cpp where the fields are stored in the database. There was already a specific situation for the RFX devices, where the extra field was excluded from the update (and insert for new devices), so I added the Netatmo type to the expression of if statement (from line 580) for the addition of new devices:
> 
> ```
>                    else if ((htype == HTYPE_RFXtrx433) || (htype == HTYPE_RFXtrx868) || (htype == HTYPE_Netatmo))
>                     {
>                             // No Extra field here, handled in CWebServer::SetRFXCOMMode and in CNetatmo::CNetatmo for Netatmo devices
>                             m_sql.safe_query("INSERT INTO Hardware (Name, Enabled, Type, LogLevel, Address, Port, SerialPort, Username, Passwo>
>                                     "DataTimeout) VALUES ('%q',%d, %d, %d,'%q',%d,'%q','%q','%q',%d,%d,%d,%d,%d,%d,%d)",
>                                     name.c_str(), (senabled == "true") ? 1 : 0, htype, iLogLevelEnabled, address.c_str(), port, sport.c_str(),>
>                                     mode2, mode3, mode4, mode5, mode6, iDataTimeout);
>                             extra = "0";
>                     }
>                     else
>                     {
> ```
> 
> and for the update from line 701 (in the original code):
> 
> ```
>                             else if ((htype == HTYPE_RFXtrx433) || (htype == HTYPE_RFXtrx868) || (htype == HTYPE_Netatmo))
>                             {
>                                     // No Extra field here, handled in CWebServer::SetRFXCOMMode and CNetatmo::CNetatmo( for Netatmo
>                                     m_sql.safe_query("UPDATE Hardware SET Name='%q', Enabled=%d, Type=%d, LogLevel=%d, Address='%q', Port=%d, >
>                                             "Mode1=%d, Mode2=%d, Mode3=%d, Mode4=%d, Mode5=%d, Mode6=%d, DataTimeout=%d WHERE (ID == '%q')",
>                                             name.c_str(), (bEnabled == true) ? 1 : 0, htype, iLogLevelEnabled, address.c_str(), port, sport.c_>
>                                             mode1, mode2, mode3, mode4, mode5, mode6, iDataTimeout, idx.c_str());
>                                     std::vector<std::vector<std::string>> result;
>                                     result = m_sql.safe_query("SELECT Extra FROM Hardware WHERE ID=%q", idx.c_str());
>                                     if (!result.empty())
>                                             extra = result[0][0];
>                             }
> ```
> 
> I implemented this change and it works as expected! :)

I have made this change also and when re-enable the plugin it keeps the authentication.
So together with `StoreRefreshToken();` in 'RefreshToken()' function. the authentication is fixed.

I'm going on holiday now, but will check later
Kind regards

**Comment by LoukKamp** [comment]
Ok, thanks and have a great vacation!

I'm still thinking about a solution for the restrictions of storing the authentication, but I think that's for another separate PR.

Maybe someone can tell me what to do to get these changes in one of the official branches, or do this for me?

**Comment by LoukKamp** [comment]
I tried to create a pull request, but failed, because I'm not a contributor, and can't find the way to become one. :(

So, please help!

**Comment by gizmocuz** [comment]
@LoukKamp , you will become one yourself once you contribute 😂

It's not that hard

- At the top of the Domoticz github project, press the 'FORK' button to make a copy on your github account
- Checkout your copy, make modifications, and commit/push your changes to it
- Now if you look on your copy github page you will see at the top a message that you can create a PR
- Once you did this, you can continue making changes to your copy if needed, the PR will be automatically updated, so no need to create a new one
- Tell me in the PR if it is ready to be merged
- Of course, test it very good

**Comment by LoukKamp** [comment]
Thanks, I'll try this this evening.

**Comment by jcbraem** [comment]
Hello,
Good news :)
I'm ready to test. When the PR is accepted, it's directly in béta chanel ?

**Comment by Phil353556** [comment]
@LoukKamp That's great,  I have done my first PR on another project a few weeks ago. 
Of course advises from @gizmocuz were good :-)  he is a master.
I'm sorry for delay .. in vacation with family

**Comment by LoukKamp** [comment]
I think it will be in the next beta, after it is accepted. Just missed the latest beta that was released today.

@Phil353556 : No problem; I thought it's better to do this myself. Not only to avoid the chance for misunderstandings, but also that should not be responsible or to blame for my mistakes (if any) ;)

**Comment by LoukKamp** [comment]
Finally, after some severe struggling with repository names and authority (access tokens), I managed to push the changes to my fork of the development repository.

It is now waiting for approval.

**Comment by LoukKamp** [comment]
The changes for improved token handling are now in the (next) beta :)

**Comment by LoukKamp** [comment]
According to Lesly from Netatmo, there is no expiration time on the refresh token:

> The refresh_token doesn't have an expiration date. It will always remain valid even if you use it several days later. Are you sure you don't have any routine sending a request to the /token endpoint that invalidates the tokens ?
> 
> You can also try - if you don't already do it - to refresh the access_token a bit before the 10800 seconds expiration to see if it can help
> 
> Have a good day,
> 
> Leslie - Community Manager
> 

So, as long as we keep the refresh token at a save place, we don't need to use the login method, based on the authorization code more than once under normal circumstances. So we don't really need to store the Client Secret and authorization code.

I expect that the PR about improving token handling will be sufficient for the moment. When this turned out to be too optimistic, a new PR, solving this issue, might be required in the future. This might be combined with a change to let the user select his/her own scope (as wished by Netatmo i.s.o. granting all devises, even those you don't use.

**Comment by LoukKamp** [comment]
@JurgenLB : As my PR includes a little change that might conflict with your (future) changes, please consider merging this change with your code.

In basics: the poll interval is defined in a compile time variable (at line 14): 


```
// 03/03/2022 - PP Changing the Weather polling from 600 to 900s. This has reduce the number of server errors, 
// 08/05/2024 - Give the poll interfval a defined name:
#define NETAMO_POLL_INTERVALL 900

```
And from line 172:

```
				//Weather, HomeCoach, and Thernistat data is updated every  NETAMO_POLL_INTERVALL  seconds 
				if ((sec_counter % NETAMO_POLL_INTERVALL == 0) || (bFirstTimeTH) || (bFirstTimeWS))
				{
					bFirstTimeWS = false;
					bFirstTimeTH = false;
					if (m_bPollWeatherData)
						GetMeterDetails();
					if (m_bPollThermostat)
						GetThermostatDetails();
				}

```

**Comment by jcbraem** [comment]
Hello,
For me this is not good. Always one error, see log after this message
Of course I have checked and re check and generate several times the ID/secret/token with full scope and weatcher scope.

I'm desesperate with this material ...
Have you any idea ?

Thanks,

Jean-Charles

2024-08-05 15:21:04.034 Status: Météo: Worker started...
2024-08-05 15:21:09.035 Météo: Requesting refreshed tokens
2024-08-05 15:21:09.398 Error: Météo: No access granted, forcing login again (Refresh tokens)...
2024-08-05 15:21:09.399 Météo: Requesting new tokens using authorization code
2024-08-05 15:21:10.414 Error: Météo: Error connecting to Server (GetMeterDetails)...
2024-08-05 15:21:10.819 Error: Météo: Error connecting to Server (GetThermostatDetails)...
2024-08-05 15:26:10.688 Error: Météo hardware (32) nothing received for more than 5 Minutes!....
2024-08-05 15:26:11.151 Error: Météo: Error connecting to Server (GetMeterDetails)...
2024-08-05 15:26:11.378 Error: Météo: Error connecting to Server (GetThermostatDetails)...
2024-08-05 15:26:11.689 Error: Restarting: Météo
2024-08-05 15:26:11.689 Status: Météo: Worker stopped...
2024-08-05 15:26:11.690 Status: Météo: Worker started...
2024-08-05 15:26:16.691 Météo: Requesting refreshed tokens
2024-08-05 15:26:17.062 Error: Météo: No access granted, forcing login again (Refresh tokens)...
2024-08-05 15:26:17.062 Météo: Requesting new tokens using authorization code
2024-08-05 15:26:17.406 Error: Météo: No access granted, check credentials (Login)...
2024-08-05 15:26:47.411 Météo: Requesting refreshed tokens
2024-08-05 15:26:47.589 Error: Météo: No access granted, forcing login again (Refresh tokens)...
2024-08-05 15:26:47.589 Météo: Requesting new tokens using authorization code
2024-08-05 15:26:47.848 Error: Météo: No access granted, check credentials (Login)...

**Comment by LoukKamp** [comment]
I know the feeling... ;)

However, you mention that you requested the authorization code, with the full scope plus weather scope (I assume a typo here), but at this moment, there is only one scope valid: see [this post](https://github.com/domoticz/domoticz/issues/6036#issuecomment-2185747817) for the only possible scope

**Comment by jcbraem** [comment]
Sorry, I have retry again with full scope and always the same issue :

2024-08-05 16:39:04.992 Status: Météo: Worker started...
2024-08-05 16:39:09.993 Météo: Requesting refreshed tokens
2024-08-05 16:39:10.222 Error: Météo: No access granted, forcing login again (Refresh tokens)...
2024-08-05 16:39:10.222 Météo: Requesting new tokens using authorization code
2024-08-05 16:39:10.610 Error: Météo: Error connecting to Server (GetMeterDetails)...
2024-08-05 16:39:10.854 Error: Météo: Error connecting to Server (GetThermostatDetails)...

**Comment by LoukKamp** [comment]
Can you show us the URL, you used to obtain the authorization code (that you will need to as the password)? Don't forget to remove your Client ID from this URL, before you post it here!

By the way: I assume that you supplied the client ID and Client Secret in the username field in the format:

```
Username: <Client ID>:<Client Secret>     e.g.: 1232323703sdfh39434hj:aa237470134jfffjj934238402
Password: <Authorization code>      e.g.: asdfghjk45676541232kjhjk
```

Note: I know this works, because I created a new device in Domoticz, using a newly created app from the Netatmo site today.
Note 2: The authorization code is only valid for a very short time (15–30 seconds?); so you have to be very quick in copying the code from the returned URL, **select the Netatmo device**, paste it in the (empty)  password field, and pressing the update button.

**Comment by LoukKamp** [comment]
> No keeping the authentication a secret means in my understanding that it is not allowed to store the tokens . The username and password are already stored.
> 
> For the scopes I was thinking of a 3e text-field where the scope can be entered. The user must then correctly fill in the scope, a selection list is ultimately better but to difficult I believe.

I wonder why we are making this so difficult? Why don't we let the user generate access and refresh tokens for his/her app at the [Netatmo app site](https://dev.netatmo.com/apps)? After specifying his/her own scope, clicking the Generate Tokens button, the user will be asked to agree with this action. When the user accepts this, the site returns to the page with the apps, where it also shows a fresh access and refresh token. We just need to have a field on the Domoticz hardware page to accept this refresh token, so the user may copy and paste this code into this field. When the user presses the update button, this token will be stored in the database, the worker will be started, and it will read this token from the database to use it to refresh the tokens again. After this, the field with the token may be cleared, so it is not shown anymore.

That's all! No need to retrieve an authorization code with complex URLs, and no need to store user credentials and the scope. This also means that the username and password fields can be used to store the client ID and client Secret without the need of combining them into one (username) field. So this will simplify the code as well.

I've tested this flow by generating a refresh token, using a scope with just my weather station. I copied the token into the database, and started the device again: **No issue at all!!**

So let me know if I can be of any help to get this done.

Edit: I was just a bit too optimistic (as usual): The combining of Client ID and secret is still required, except when we add an extra field to handle the token. But I think that can be done, and I'm willing to do that.

**Comment by JurgenLB** [comment]
I think this is a question for @kiddigital why he chose for this authentication method.

But I think it is because you need extra fields for the tokens.

You can also chose for a extra field for the scopes.

**Comment by jcbraem** [comment]
Hello,
After many try I have found one solution : need to create a new app into Netatmo dev portal.
Directly OK !
After one restart of Domoticz I have a new token and all informations are retrieve from Netatmo.
Thank you for your job :)

**Comment by LoukKamp** [comment]
> Hello, After many try I have found one solution : need to create a new app into Netatmo dev portal. Directly OK ! After one restart of Domoticz I have a new token and all informations are retrieve from Netatmo. Thank you for your job :)

OK, great that you found this so fast; It took me more than a week to find that the app I was using for so long, became cripple!

It's good to mention here, but also on the forum, that this may happen: When you are not able to create valid tokens, and you are sure you did it all the right way, creating a new app on the Netatmo device portal might be the solution!

**Comment by LoukKamp** [comment]
> I think this is a question for @kiddigital why he chose for this authentication method.
> But I think it is because you need extra fields for the tokens.

OK, so I'm curious about @kiddigital his thoughts about this.

> 
> You can also chose for a extra field for the scopes.

Why? The scope is only needed in the request for the authorization code. When we skip this phase, we don't need a field or a storage for this scope, and/or this authorization code; just the refresh token. 

So, we can keep it this way (combining client ID and secret in the username field, and use the password field for the refresh token), or we create an extra field for this and make the user's life a little bit less complex.

**Comment by kiddigital** [comment]
> I think this is a question for @kiddigital why he chose for this authentication method.
> 
> But I think it is because you need extra fields for the tokens.
> 
> You can also chose for a extra field for the scopes.

Hi, I did not really choose anything specific but if you look at the discussion of issue #5766, the only thing I did was _fix_ the old broken API key and replaced it with the (Oauth2) _authorization_code_, so proper Access- and Refresh Tokens can be retrieved. (As I am not an owner/user of any Netatmo hardware, I also don't use the Portals ;) )

And yes, it could be an option that instead of entering the _authorization_code_, we could let the users enter the Refresh Token and go from there. But that only works for those Portals, like the current Netatmo Dev Portal apparently, where the option is provided to generate these tokens. It it not a common pattern to do that. But retrieving an _authorization_code_ from an OAuth2 flow is.

Good to see that others, including actual Netatmo hardware owners, are contributing to keep this hardware supported.

**Comment by LoukKamp** [comment]
Thanks for this response. So, we have several options, all with there own advantages and disadvantages:

1. Just change the workflow: The user needs to fill in a refresh token I s.o. the authorization code. The first valid refresh token can be generated om the device portal on the same page as where the user creates the app, and where the the client credentials are shown.
2. We implement a login procedure, where the user is lead to the user logs in, and accepts the conditions. The returned authorization code is read from the response, and used to obtain the tokens.
3. Optional: We add a field on the hardware page for the Netatmo device to input the authorization code or the refresh token. This avoid complex input handling for the user where he/she needs to combine cliënt ID's and secrets.

The first option is the most simple one. However, this might not work with other devices we support and use oAuth, without the option to generate tokens.

When we want to have one standard methode for all those devices, we need to implement option 2, maybe combined with option 3.

Any other thoughts?

**Comment by gizmocuz** [comment]
I vote for simplicity for the end-user... 'option 2', logging-in via a webportal to create a token seems complicated, and the user already know it's username/password

**Comment by JurgenLB** [comment]
I believe that if you use the portal and the generated tokens you must have a field for the authentication token and a other field for the refresh token.
But like @gizmocuz mentioned the automated login with use of retrieving an authorization_code from an OAuth2 flow is the preferred way.

In this case we need a third field for the scopes I believe. We now have the username containing ClientID and ClientSecret.
This is also not a user memorisable setup.

**Comment by gizmocuz** [comment]
I'm not sure about the Scope selection, makes it also complicated for end-users. What scope should be picked?
Maybe some checkboxes that the user can enable to specify what hardware is in use? Temperature, thermostat, etc.

**Comment by Phil353556** [comment]
@gizmocuz I fully agree with your idea:  checkboxes.
see https://dev.netatmo.com/apidocumentation/oauth#scopes

Access tokens are associated to a scope, when requesting one, you should precise which scope you require. There are nine available scopes:
-read_station: to retrieve weather station data ([Getstationsdata](https://dev.netatmo.com/apidocumentation/weather#getsationdata), [Getmeasure](https://dev.netatmo.com/apidocumentation/weather#getmeasure))
-read_thermostat: to retrieve thermostat data ( [Homestatus](https://dev.netatmo.com/apidocumentation/energy#homestatus), [Getroommeasure](https://dev.netatmo.com/apidocumentation/energy#getroommeasure)...)
-write_thermostat: to set up the thermostat ([Synchomeschedule](https://dev.netatmo.com/apidocumentation/energy#synchomeschedule), [Setroomthermpoint](https://dev.netatmo.com/apidocumentation/energy#setroomthermpoint)...)
-read_camera: to retrieve Smart Indoor Cameradata ([Gethomedata](https://dev.netatmo.com/apidocumentation/security#gethomedata), [Getcamerapicture](https://dev.netatmo.com/apidocumentation/security#getcamerapicture)...)
-write_camera: to inform the Smart Indoor Camera that a specific person or everybody has left the Home ([Setpersonsaway](https://dev.netatmo.com/apidocumentation/security#setpersonsaway), [Setpersonshome](https://dev.netatmo.com/apidocumentation/security#setpersonshome))
-access_camera: to access the camera, the videos and the live stream *
-read_presence: to retrieve Smart Outdoor Camera data ([Gethomedata](https://dev.netatmo.com/apidocumentation/security#gethomedata), [Getcamerapicture](https://dev.netatmo.com/apidocumentation/security#getcamerapicture)...)
-access_presence: to access the camera, the videos and the live stream *
-read_smokedetector : to retrieve the Smart Smoke Alarm informations and events ([Gethomedata](https://dev.netatmo.com/apidocumentation/security#gethomedata), [Geteventsuntil](https://dev.netatmo.com/apidocumentation/security#geteventsuntil)...)
-read_homecoach: to read data coming from Smart Indoor Air Quality Monitor ([gethomecoachsdata](https://dev.netatmo.com/apidocumentation/aircare#gethomecoachsdata))
If no scope is provided during the token request, the default is "read_station"

**Comment by LoukKamp** [comment]
Thanks for the research; maybe we can show the detailed information somewhere to assist the user.

I already tried some things in the HTML/jd code, but had very less time to work it out; maybe more this weekend...

**Comment by JurgenLB** [comment]
>available scopes: -read_station: to retrieve weather station data ([Getstationsdata](https://dev.netatmo.com/apidocumentation/weather#getsationdata), [Getmeasure](https://dev.netatmo.com/apidocumentation/weather#getmeasure)) -read_thermostat: to retrieve thermostat data ( [Homestatus](https://dev.netatmo.com/apidocumentation/energy#homestatus), [Getroommeasure](https://dev.netatmo.com/apidocumentation/energy#getroommeasure)...) -write_thermostat: to set up the thermostat ([Synchomeschedule](https://dev.netatmo.com/apidocumentation/energy#synchomeschedule), [Setroomthermpoint](https://dev.netatmo.com/apidocumentation/energy#setroomthermpoint)...) -read_camera: to retrieve Smart Indoor Cameradata ([Gethomedata](https://dev.netatmo.com/apidocumentation/security#gethomedata), [Getcamerapicture](https://dev.netatmo.com/apidocumentation/security#getcamerapicture)...) -write_camera: to inform the Smart Indoor Camera that a specific person or everybody has left the Home ([Setpersonsaway](https://dev.netatmo.com/apidocumentation/security#setpersonsaway), [Setpersonshome](https://dev.netatmo.com/apidocumentation/security#setpersonshome)) -access_camera: to access the camera, the videos and the live stream * -read_presence: to retrieve Smart Outdoor Camera data ([Gethomedata](https://dev.netatmo.com/apidocumentation/security#gethomedata), [Getcamerapicture](https://dev.netatmo.com/apidocumentation/security#getcamerapicture)...) -access_presence: to access the camera, the videos and the live stream * -read_smokedetector : to retrieve the Smart Smoke Alarm informations and events ([Gethomedata](https://dev.netatmo.com/apidocumentation/security#gethomedata), [Geteventsuntil](https://dev.netatmo.com/apidocumentation/security#geteventsuntil)...) -read_homecoach: to read data coming from Smart Indoor Air Quality Monitor ([gethomecoachsdata](https://dev.netatmo.com/apidocumentation/aircare#gethomecoachsdata)) If no scope is provided during the token request, the default is "read_station"

This is from the OLD API,

for the NEW API the scopes are;
Weather: [read_station]
Home + Control (Magellan): [read_magellan] [write_magellan]
BubenDorff: [read_bubendorff] [write_bubendorff]
Smart Thermostat: [read_smarther] [write_smarther]
Bticino Class 300 EOS: [read_mx] [write_mx]
My Home Server 1 : [read_mhs1] [write_mhs1]
Netatmo Thermostat/valves: [read_thermostat] [write_thermostat]
Indoor Camera / Doortags / indoor Sirene: [read_camera] [write_camera] [access_camera]
Doorbell: [read_doorbell] [access_doorbell]
Outdoor Camera: [read_precense] [write_precense] [access_presence]
Home Coach : [read_homecoach]
CO2 : [read_carbonmonoxidedetector]
Smoke : [read_smokedetector]

kind regards

**Comment by Phil353556** [comment]
ok
source of what you share is: https://dev.netatmo.com/apidocumentation/general
and it includes third parties of netatmo or compatible devices
kind regards

**Comment by kiddigital** [comment]
Why not just use all available scopes in the authorization code request?

I don't think that Netatmo has a sophisticated system running that validates if the requested scopes are logical options given all the hardware that one has...

1 note on 'option 2' (login at Netatmo to get the authorization code). Normally that would require that a proper HTTP(S) redirect Uri is configured when setting up the 'client' application (on the Netatmo dev portal). For the Domoticz instance of the user that needs the code, what should the User fill in? The domoticz instance runs in ones home network. The device that runs the browser that the User uses to access Domoticz (UI) is also the one that has to handle the redirect Uri.

There are a lot of ways to go about this, but is requires good network (resolving, etc.) understanding (and skills probably).

One of the best options, at least in my opinion, would be if Domoticz would support mDNS. This way, each end user device (browser or app), that is on the same network as the domoticz instance, can request for example 'domoticz.local'. No need to configure anything on the local network (like DNS) as domoticz would listen for these requests and answer accordingly.

This would resolve the issue for OAuth2/OIDC authorization code flows (and others) as the redirect Uri could be known (domoticz.local).

For those interested, a while back a started a branch (still WIP) on my github development fork.

**Comment by JurgenLB** [comment]
> ok source of what you share is: https://dev.netatmo.com/apidocumentation/general and it includes third parties of netatmo or compatible devices kind regards

Yes it is mentioned there but I also research the returning API, the third Party I mentioned for being thorough.

@kiddigital I dit try using all the scopes but that failed. Because I don't have a doorbell so I deleted the Doorbell from my code and It returned to work.
I do like your idea from domoticz.local but that is for the further then.

kind regards

**Comment by kiddigital** [comment]
> 
> @kiddigital I dit try using all the scopes but that failed. Because I don't have a doorbell so I deleted the Doorbell from my code and It returned to work.
> 

Ok, well done Netatmo.
So that means if domoticz would make it user friendly and offer checkboxes, these checkboxes should not show/request 'scopes' but should rather show all possible (supported) Netatmo devices and domoticz has to 'derive' which of the scopes of the overall list matches the total set of scopes needed for all selected (owned) devices.

**Comment by JurgenLB** [comment]
> > @kiddigital I dit try using all the scopes but that failed. Because I don't have a doorbell so I deleted the Doorbell from my code and It returned to work.
> 
> Ok, well done Netatmo. So that means if domoticz would make it user friendly and offer checkboxes, these checkboxes should not show/request 'scopes' but should rather show all possible (supported) Netatmo devices and domoticz has to 'derive' which of the scopes of the overall list matches the total set of scopes needed for all selected (owned) devices.

sorry but I'm no Professional but that is what happened with my setup.
I do hope we can have more Netatmo devices supported in Domoticz.
As for the supported devices that's why I mentioned the whole list of scopes with the suggested names.

But apparently this is not true for all devices because we now have users whiteout a smoke detector and the Old API integration works.

kind regards

**Comment by LoukKamp** [comment]
Someone mentioned that the Domoticz development team seems to have some kind of agreement: Do not store user credentials, and keep the scope as small as possible. So, that's the reason why we need to offer the user to select his own scope, and should not fix the scope to the full range.  I don't know the reason behind this, but I can guess that it has something to do with the costs of data traffic and server capacity.

In the meanwhile, I'm scanning the possibilities tho automate the login sequence. The trouble is that the user is requested to fill in some credentials and needs to accept the conditions, but we need to intercept the response in a procedure to store the result (the authorization code).

Because I'm quite inexperienced in HTML front end processing, this evaluation process is quite slow in my case, so please have some patience with me...

A sample of what I think it should look like:

![image](https://github.com/user-attachments/assets/dada73ac-7b4a-4a14-9a2e-2b17726a2367)

Note: In the example above, there are 2 options selected for the scope (highlighted). For me, this is sufficient, but I guess, some people prefer select boxes instead. However, I did not find an easy way to add support for select boxes without a lot of extra coding in JavaScript so far.

**Comment by LoukKamp** [comment]
So far, no luck :(

I try to use XMLHttpRequest to get logged in, but I never get a response from the servers at Netatmo. Using Firefox results in some CORS exceptions, and when using Chrome, I just get no response at all. I think, this has something to do with redirection to the auth.netatmo.com site. Also tried to address the auth site directly, but also no response. When using postman, the response is always a login page in HTML, even when I add a header with "access = application/JSON".

Now I'm also wondering why I'm trying this?  To login, we need the client ID of the app (next to the username and password). So the user needs to log in on the device portal to get the client ID of the app you want to use (there may be more than one app, each with its own client ID). So, when the user may also generate the refresh token at that time and copy that into the corresponding field on the Domoticz hardware page for the Netatmo device.

Any hints or suggestions, where to look for alternatives?

**Comment by LoukKamp** [comment]
Finally, a breakthrough!!!

Now, I'm using a windows.open method to contact the Netatmo site for authorization. There are still CORS exceptions, but only when the window displays data from the Netatmo sites, but as soon as it is redirected to the (local) Domoticz site, the responses aren't blocked anymore by CORS and I can extract the returned authorization code from the response.

The trick was to find an address within the range of valid addresses of the (local) Domoticz server that won't lead to an exception: As soon as I get a valid response from Domoticz, the response is not blocked by CORS anymore...

The challenge was/is to find an entrance (URL) that is accepted by the Domoticz server without the presence of a hash-sign (#) ; URLs that contains a hash-sign in the redirect_uri, will be corrupted by the Netatmo servers somewhere in the process.  The only valid redirect_uri is something like "http://domoticzserver:7184/" where domoticzserver is the hostname of my PI, running Domoticz. It can't be localhost, like suggested everywhere, because localhost is the system where your browser is running on.

It needs quite a bit of ironing and maybe there can be a special URL (without the hash-sign) for this purpose (http://domoticsz:7184/authorize/Netatmo for instance).

I will update the changes to my own fork asap, so you can see what is changed.

**Comment by JurgenLB** [comment]
Is there no variable from Domoticz that contains the IP-adress/port?
So you do `http://192.168.1.1:8080` as redirect_uri. 
When Domoticz is started there is in the Log `Status: WebServer(HTTP) started on address: :: with port 8080`
so I think there is some variable available but in the log it is not showing correctly.

For the scopes I think this is a correct example

![Netatmo Scopes](https://github.com/user-attachments/assets/feff64f7-2e87-426f-afb3-15972b17c970)

only the functions for the values need to be adapted.

**Comment by LoukKamp** [comment]
I was thinking of using the URL of the current (hardware page), but still work in progress...

Personally, I would prefer a (scrollable) select list, like the example I gave some posts again (and used this for my last changes).

**Comment by JurgenLB** [comment]
> I was thinking of using the URL of the current (hardware page), but still work in progress...
> 
> Personally, I would prefer a (scrollable) select list, like the example I gave some posts again (and used this for my last changes).

But is the (scrollable) select list compatible with all the browsers and OS-systems for multiple selection?
I believe it is not, so the checkbox is more compatible.

Kind regards

**Comment by LoukKamp** [comment]
According to W3 schools, it is (see [this page](https://www.w3schools.com/tags/att_select_multiple.asp))

**Comment by LoukKamp** [comment]
Finally, after a week of struggling, I managed to get it working in JavaScript running in the front-end environment (browser). Why? Because the first part (acquiring the authorization code) needs user response / interaction, and this can't be done from the back-end. The second part (exchanging the authorization code for access and refresh tokens) needs to be done within a view seconds after the code is received (very short expiration time). And running the first in the front-end, and the second in the back-end, needed too many variables to transport (via the database) from one to another.

The main issue, I was struggling with, was an "Invalid or Missing grant-type parameter". In the end, it turned out that the parameters should be supplied in the body of the request and not as URL-parameters, as the documentation suggests. Finally, I managed to figure this out after testing with postman, making network traces, and analyzing those with Wireshark (not easy when encryption is used!). Also, an authorization code may be used only once, even if the call failed.

Now, I continue ironing, and I hope this will not lead me into trouble again... ;)

**Comment by jcbraem** [comment]
LoukKamp,

Many thanks for your "relentlessness" !

**Comment by LoukKamp** [comment]
Thanks, I learned a new word again 😉Maybe this is why they called me the pitbull at work...😁LoukOp 19 aug. 2024 07:29 schreef jcbraem ***@***.***>:
LoukKamp,
Many thanks for your "relentlessness" !

**Comment by sidepipe** [comment]
Things seem to be broken somehow ( at least for me. ) When I click the "Login" button to retrieve the access/refresh tokens, I get "Access denied: Failed to receive a valid reponse from server (400): error = invalid_grant"
Thing is, I'm not sure that this is specific to Domoticz... I'm also using an MQTT bridge which has suddenly started giving the same error after months of working fine. Any ideas what's going on?

**Comment by LoukKamp** [comment]
I don't  know about the MQTT bridge, but the Invalid Grant message is probably caused by having some text in the Redirect URI field on the Netatmo device portal: Clear this field and try to log in again (by clicking the login button on the Netatmo device page).

**Comment by JurgenLB** [comment]
> Things seem to be broken somehow ( at least for me. ) When I click the "Login" button to retrieve the access/refresh tokens, I get "Access denied: Failed to receive a valid reponse from server (400): error = invalid_grant"
> 
> Thing is, I'm not sure that this is specific to Domoticz... I'm also using an MQTT bridge which has suddenly started giving the same error after months of working fine. Any ideas what's going on?

Because you mention an MQTT-bridge I suspect you use the same credentials in the 2 software program's. You must create 2 apps in the dev.netatmo cloud and never use the same app-credentials on a second system, always make a new app.

**Comment by waltervl** [comment]
Closed, see issue 6280

**Comment by PvdGulik** [comment]
Hello, I just started to use the latest update to 2025.1, hoping it would solve the problems with Netatmo.
But still getting the error's in the logging..

> 2025-05-14 14:19:28.712 Error: Netatmo: Refresh token false 0
> 2025-05-14 14:19:30.072 Error: Netatmo: No access granted, forcing login again (Refresh tokens): invalid_grant
> 2025-05-14 14:19:30.073 Error: Netatmo: Refresh token false 0
> 2025-05-14 14:19:31.435 Error: Netatmo: No access granted, forcing login again (Refresh tokens): invalid_grant
> 2025-05-14 14:19:31.436 Error: Netatmo: Refresh token false 0

I followed this instruction, without errors, but still the errors occur in the logging.

What else to do?

Version: 2025.1
Build Hash: 89d5c900d
Compile Date: 2025-05-05 09:02:49
dzVents Version: 3.1.8
Python Version: 3.11.2 (main, Nov 30 2024, 21:22:50) [GCC 12.2.0]

Thanks, Pieter

**Comment by Phil353556** [comment]
Hello @PvdGulik,

Got the same issue, with the same errors messages
look at https://forum.domoticz.com/viewtopic.php?p=325691
Redirect URL must be blanked
now it works
Phil

**Comment by PvdGulik** [comment]
It was not the solution for me. Because I have Domoticz and Homebridge running on the same raspberry, I tried to use the client-ID and clientsecret of homebridge in Domoticz, and that seemed to help...

In the hombridge app the Redirect URL was blank as well.

Thanks!

**Comment by JurgenLB** [comment]
Homebridge and Domoticz can run on the same system without problems. The Netatmo plugin has no relationship with homebridge.

If you use a app credentials from Netatmo for Homebridge,
Than you must use a different app credentials from Netatmo for Domoticz.

Kind Regards

**Comment by PvdGulik** [comment]
Strange, for me it is working now for the first time, since march 2025…
I’m happy now
